### PR TITLE
Errno fix

### PR DESCRIPTION
**Pull request by rmn30**
Here is a fix for #461 and the associated test. I've done separate commits for the test and the bug fix. Not sure if you would prefer them to be squashed?

### CODE REVIEW

**Comment by rmn30** [comment]
After a bit of tinkering and some more thought I am withdrawing this PR because I think this is not the correct solution: snmalloc should not be setting `errno` except for functions that are defined to do so probably the override wrapper is the correct place for this (at the cost of a `nullptr` check).

**Comment by mjp41** [comment]
I don't want a branch on the fast path of `malloc`, and setting `errno` at the outer level will lead to this.  It has to be buried deep inside.

**Comment by rmn30** [comment]
In that case setting `errno` should become part of snmalloc's API and we will have to save and restore `errno` in the `posix_memalign` implementation because it explicitly does not set it. Implementations of the other functions already check for `alloc` returning null and set `errno`, although the `realloc` implementation is a bit broken and I will fix it.

**Comment by mjp41** [comment]
Does `posix_memalign` require it not to be set, or that you can't depend on it being set?

In particular, I read 
https://sourceware.org/bugzilla/show_bug.cgi?id=83#c1

Which implies we can set it.

**Comment by davidchisnall** [comment]
POSIX2018 is [pretty liberal WRT `errno`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/errno.html):

 - It is never set to 0 by any POSIX function.
 - It is undefined after any call to a POSIX function that is not explicitly stated to set it.
 - It is undefined after a successful call to a POSIX function that is documented to set `errno` on failure.

[`posix_memalign`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/posix_memalign.html) is not defined to set `errno`, so we're free to set or not set it as we desire on failure.  On success we may set it to whatever we want as long as we do not transition it from a non-zero to a zero state.

**Comment by mjp41** [comment]
To get this right, some of the other pals should probably set it.  Like 
https://github.com/microsoft/snmalloc/blob/af8ab2daf6455fa1f9d9daedeb65f6abc2d2d50e/src/pal/pal_windows.h#L163-L196

Windows Pal should probably set `errno` when `VirtualAlloc*` returns `nullptr`.

File: src/test/func/malloc/malloc.cc [hunk 1]
```diff
@@ -14,10 +14,10 @@ constexpr int SUCCESS = 0;
 void check_result(size_t size, size_t align, void* p, int err, bool null)
 {
   bool failed = false;
-  if (errno != err && err != SUCCESS)
+  if (errno != err)
```
[the remark this answers is not in the document]
Comment (rmn30) [reply src/test/func/malloc/malloc.cc:17]: I haven't seen this so far.

**Review by mjp41 — APPROVED** [review]
LGTM, thanks for fixing

**Pull request merged**